### PR TITLE
Enhancement: Support for SHA-256 file checksums

### PR DESCRIPTION
**Pull request by elliotweiser**
Summary
-------

This change adds support for SHA-256 file checksums. Example:

```yaml
file:
  /goss/hellogoss.txt:
    exists: true
    sha256: 7f78ce27859049f725936f7b52c6e25d774012947d915e7b394402cfceb70c4c
```

This change includes:

* Go code (modeled from md5 functionality) for a sha256 equivalent
* A new (shared) integration test for `sha256` file checksums.
* Documentation update showcasing a `sha256` example.

Additional changes
------------------

I had to make some additional changes because `make deps && make` (as per
the `.travis.yml`) didn't work as is. There seemed to be changes in either
the Docker base images or in upstream package repos (e.g. a specific `apache2`
package version was just _gone_). I also noticed that the ArchLinux container
fails to start properly when invoked via the `test.sh` script because `/sbin/init`
doesn't exist. After digging through some community forums and stackoverflow posts
I didn't understand, I decided it wasn't worth the effort to "fix", so I replaced
the command with a `sleep 1h`. Fortunately, the container still gets cleaned up
and it seems to work fine.

To summarize:

* Integration test/Dockerfile updates
* ArchLinux Docker image doesn't have `/sbin/init` (anymore)

### CODE REVIEW

**Comment by aelsabbahy** [comment]
Looks good, can you try to lock the apache version down to the new one.

It's a bit hackish, but I use the md5 of the dockerfiles to cache them and speed up the build.

Strange about arch, I can investigate further later..

**Comment by elliotweiser** [comment]
Pinned the version to Apache version on wheezy to `2.2.22-13+deb7u8`. I actually had to make even more changes originally to get all of the `make` targets to work on my Mac (using Docker for Mac). Apparently some of those issues don't come up in the CI environment 🤷‍♂️ .

In any case, let me know if you need anything else to get this merged.

**Comment by elliotweiser** [comment]
My observation about ArchLinux may have only surfaced when attempting to run said container on a Mac. I'll at least see if that part of this PR can be reverted...

**Comment by elliotweiser** [comment]
Yep works in Travis CI 👍

Ready when you are @aelsabbahy 😄

**Comment by elliotweiser** [comment]
@aelsabbahy Any updates on getting this reviewed/merged?

**Comment by aelsabbahy** [comment]
I'll try to get to it this week. Sorry for the delay.

**Comment by elliotweiser** [comment]
@aelsabbahy Just checking in. Any progress?

**Comment by elliotweiser** [comment]
@aelsabbahy Rebased and resolved conflicts 👍

**Comment by elliotweiser** [comment]
@aelsabbahy I shook out the regressions from Travis CI. Should be all fixed now. 

Since I had to update the apache version again on wheezy, I opened #253 for your consideration.

**Comment by aelsabbahy** [comment]
Can you try reverting the `Dockerfile` to the one on master. I think That will maintain the old version and would make all the version changes in the test files unnecessary.

See:
https://travis-ci.org/aelsabbahy/goss/builds/250023452#L1961

**Comment by elliotweiser** [comment]
Sure. Give me a few minutes...

**Comment by elliotweiser** [comment]
I did as you asked, but it looks like a recent change introduced a somewhat brittle unit test (I observed that error in a few upstream Travis CI builds). I could amend and force-push until it "works", but it would probably make more sense to address the failing test first.

**Comment by aelsabbahy** [comment]
@OmarDarwish it appears the have-key-with-value test is brittle. Can you take a look at it when you get a chance?

See: https://travis-ci.org/aelsabbahy/goss/builds/250194911

This is the second time I've seen it failing in the past 3 days or so.

**Comment by elliotweiser** [comment]
I just reproduced it locally. If you run `make test` enough times, it shows up.

```
--- FAIL: TestMatcherToGomegaMatcher (0.00s)
	gomega_test.go:145: For input '{"have-key-with-value": { "foo": 1, "bar": "baz" }}': got *matchers.AndMatcher &{[0xc4202085a0 0xc4202085c0] <nil>}, want *matchers.AndMatcher &{[0xc420114f60 0xc420114f80] <nil>}
FAIL
FAIL	github.com/aelsabbahy/goss/resource	0.012s
ok  	github.com/aelsabbahy/goss/system	0.011s
?   	github.com/aelsabbahy/goss/util	[no test files]
?   	github.com/aelsabbahy/goss	[no test files]
make: *** [test] Error 1
```

**Comment by aelsabbahy** [comment]
Probably has to do with golang maps being randomized. That's just a guess without digging into the code yet.

**Comment by OmarDarwish** [comment]
@aelsabbahy that's my thought as well. Will dig into this today as soon as I can. My bad guys!

**Comment by elliotweiser** [comment]
@OmarDarwish No worries. These things happen. Thanks for looking into it! 👍

**Comment by aelsabbahy** [comment]
No worries, thanks for the quick response and contributions.

**Comment by OmarDarwish** [comment]
#256 should fix flakiness
